File: lib/Screen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Point_1 = require("./Point");
const util_1 = require("./util");
class Screen {
    constructor(width, height, map, shapes = []) {
        this.width = width;
        this.height = height;
        this.point_mapping = map;
        this.shapes = shapes;
    }
    offset() {
        return new Point_1.default(0, 0);
    }
    occupiedArea() {
        const { width, height } = this;
        const matrix = util_1.buildMatrix(width, height);
        this.shapes.forEach((shape) => {
            const { x, y } = shape.offset();
            const shape_width = shape.width;
            const shape_height = shape.height;
            shape.occupiedArea().forEach((row, dy) => {
                row.forEach((value, dx) => {
                    const mapped = this.point_mapping(new Point_1.default(x + dx, y + dy), width, height, shape_width, shape_height);
                    matrix[mapped.y][mapped.x] += value;
                });
            });
        });
        return matrix;
    }
    addShape(shape) {
        this.shapes.push(shape);
    }
}
exports.default = Screen;

File: src/Screen.py
from src.Point import Point
from src.Shape import Shape
from src.util import buildMatrix

class Screen(Shape):
  def __init__(self, width, height, map, shapes):
    self.width = width
    self.height = height
    self._point_mapping = map

    # putting a list as default shares the pointer to that list
    # among all instances
    if shapes is None:
      self._shapes = []
    else:
      self._shapes = shapes

  def offset(self):
    return Point(0, 0)

  def occupiedArea(self):
    width = self.width
    height = self.height
    matrix = buildMatrix(width, height)

    for shape in self._shapes:
      offset = shape.offset()
      x = offset.x
      y = offset.y
      shape_width = shape.width
      shape_height = shape.height

      for dy, row in enumerate(shape.occupiedArea()):
        for dx, value in enumerate(row):
          mapped = self._point_mapping(
            Point(x + dx, y + dy),
            width, height,
            shape_width, shape_height,
          )

          matrix[mapped.y][mapped.x] += value

    return matrix

  def addShape(self, shape):
    self._shapes.append(shape)
